compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -336,75 +308,83 @@ public CodeAnalyzer(CompilerContext context) {
         this.names = Names.getInstance(context);
         this.symResolver = SymbolResolver.getInstance(context);
         this.reachabilityAnalyzer = ReachabilityAnalyzer.getInstance(context);
-        this.enableExperimentalFeatures = Boolean.parseBoolean(
-                CompilerOptions.getInstance(context).get(CompilerOptionName.EXPERIMENTAL));
     }
 
+    // We need this `analyze`
     public BLangPackage analyze(BLangPackage pkgNode) {
-        this.workerReferences.clear();
-        this.dlog.setCurrentPackageId(pkgNode.packageID);
-        pkgNode.accept(this);
+        final AnalyzerData data = new AnalyzerData();
+        visitNode(pkgNode, data);
         return pkgNode;
     }
 
     @Override
-    public void visit(BLangPackage pkgNode) {
+    public void visit(BLangPackage pkgNode, AnalyzerData data) {
+        this.dlog.setCurrentPackageId(pkgNode.packageID);
         if (pkgNode.completedPhases.contains(CompilerPhase.CODE_ANALYZE)) {
             return;
         }
-        parent = pkgNode;
-        SymbolEnv pkgEnv = this.symTable.pkgEnvMap.get(pkgNode.symbol);
-        analyzeTopLevelNodes(pkgNode, pkgEnv);
-        pkgNode.getTestablePkgs().forEach(testablePackage -> visit((BLangPackage) testablePackage));
+        data.parent = pkgNode;
+        data.env = this.symTable.pkgEnvMap.get(pkgNode.symbol);
+        analyzeTopLevelNodes(pkgNode, data);
+        pkgNode.getTestablePkgs().forEach(testablePackage -> visitNode(testablePackage, data));
     }
 
-    private void analyzeTopLevelNodes(BLangPackage pkgNode, SymbolEnv pkgEnv) {
+    private void analyzeTopLevelNodes(BLangPackage pkgNode, AnalyzerData data) {
         List<TopLevelNode> topLevelNodes = pkgNode.topLevelNodes;
-        for (int i = 0; i < topLevelNodes.size(); i++) {
-            TopLevelNode topLevelNode = topLevelNodes.get(i);
-            analyzeNode((BLangNode) topLevelNode, pkgEnv);
+        for (TopLevelNode topLevelNode : topLevelNodes) {
+            analyzeNodex((BLangNode) topLevelNode, data);
         }
         pkgNode.completedPhases.add(CompilerPhase.CODE_ANALYZE);
-        parent = null;
     }
 
-    private void analyzeNode(BLangNode node, SymbolEnv env) {
-        SymbolEnv prevEnv = this.env;
-        this.env = env;
-        BLangNode myParent = parent;
+    // TODO: Change the method name
+    private void analyzeNodex(BLangNode node, AnalyzerData data) {
+        SymbolEnv prevEnv = data.env;
+        analyzeNodeWithParent(node, data);
+        data.env = prevEnv;
+    }
```

Comment:
Originally this is a private method and we have an abstract method in `SimpleBLangNodeAnalyzer`
```java
public abstract void analyzeNode(BLangNode node, T data);
```
@hasithaa, is that okay to make this method public? or we can rename this method to avoid conflict.